# D212 - Data Mining II
### OFM3 Task 3: Association Rules and Lift Analysis
#### Data Mining II — D212
#### PRFA — OFM3
> André Davis
> StudentID: 010630641
> MSDA
>
> Competencies
> 4030.06.6 : Pattern Prediction
> The graduate predicts patterns in data using association rules and lift analysis.

#### Table of Contents
<ul>
    <li><a href="#data-set-medications">Dataset Medications</a></li>
    <li><a href="#research-question">A1: Research Question</a></li>
    <li><a href="#objectives">A2: Objectives and Goals of Analysis</a></li>
    <li><a href="#explination-of-market-basket-analysis">B1: Explanation of Market Basket Analysis</a></li>
    <li><a href="#transaction-example">B2: Transaction Example</a></li>
    <li><a href="#assumptions">B3: Assumptions of Market Basket Analysis</a></li>
    <li><a href="#data-prepartion">C1: Data Preparation</a></li>
    <li><a href="#geneartion-of-association-rules">C2: Generation of Association Rules</a></li>
    <li><a href="#association-rules-table">C3: Association Rules Table</a></li>
    <li><a href="#top-rules">C4: Top Rules</a></li>
    <li><a href="#results-analysis">D1: Results of Analysis</a></li>
    <li><a href="#practical-significance">D2: Practical Significance</a></li>
    <li><a href="#recommended-action">D3: Recommended Action</a></li>
    <li><a href="#panopto-recording">E: Panopto Recording</a></li>
    <li><a href="#code-references">F: Code References</a></li>
    <li><a href="#source-references">G: Source References</a></li>
</ul>

<a id="data-set-medications"></a>
# Dataset Medications 
#### - Additional Information, non-required Section

##### Medication Details to help with understanding the association rules.

| Medication Name            | Treats                            | Source Citation                                                                                          |
|----------------------------|-----------------------------------|----------------------------------------------------------------------------------------------------------|
| Abilify                    | Mental health disorders           | [Abilify - Wikipedia](https://en.wikipedia.org/wiki/Aripiprazole)                                        |
| Acetaminophen              | Pain and fever relief             | [Acetaminophen - Wikipedia](https://en.wikipedia.org/wiki/Acetaminophen)                                 |
| Actonel                    | Osteoporosis                      | [Risedronic acid - Wikipedia](https://en.wikipedia.org/wiki/Risedronic_acid)                             |
| Albuterol Aerosol          | Asthma and bronchospasm           | [Albuterol - Wikipedia](https://en.wikipedia.org/wiki/Albuterol)                                         |
| Albuterol HFA              | Asthma and bronchospasm           | [Albuterol - Wikipedia](https://en.wikipedia.org/wiki/Albuterol)                                         |
| Alendronate                | Osteoporosis                      | [Alendronic acid - Wikipedia](https://en.wikipedia.org/wiki/Alendronic_acid)                             |
| Allopurinol                | Gout                              | [Allopurinol - Wikipedia](https://en.wikipedia.org/wiki/Allopurinol)                                     |
| Alprazolam                 | Anxiety and panic disorders       | [Alprazolam - Wikipedia](https://en.wikipedia.org/wiki/Alprazolam)                                       |
| Amitriptyline              | Depression and neuropathy         | [Amitriptyline - Wikipedia](https://en.wikipedia.org/wiki/Amitriptyline)                                 |
| Amlodipine                 | Hypertension                      | [Amlodipine - Wikipedia](https://en.wikipedia.org/wiki/Amlodipine)                                       |
| Amoxicillin                | Bacterial infections              | [Amoxicillin - Wikipedia](https://en.wikipedia.org/wiki/Amoxicillin)                                     |
| Amphetamine                | ADHD and narcolepsy               | [Amphetamine - Wikipedia](https://en.wikipedia.org/wiki/Amphetamine)                                     |
| Amphetamine Salt Combo     | ADHD                              | [Amphetamine/dextroamphetamine - Wikipedia](https://en.wikipedia.org/wiki/Amphetamine/dextroamphetamine) |
| Amphetamine Salt Combo XR  | ADHD                              | [Amphetamine/dextroamphetamine - Wikipedia](https://en.wikipedia.org/wiki/Amphetamine/dextroamphetamine) |
| Atenolol                   | Hypertension                      | [Atenolol - Wikipedia](https://en.wikipedia.org/wiki/Atenolol)                                           |
| Atorvastatin               | High cholesterol                  | [Atorvastatin - Wikipedia](https://en.wikipedia.org/wiki/Atorvastatin)                                   |
| Azithromycin               | Bacterial infections              | [Azithromycin - Wikipedia](https://en.wikipedia.org/wiki/Azithromycin)                                   |
| Benazepril                 | Hypertension                      | [Benazepril - Wikipedia](https://en.wikipedia.org/wiki/Benazepril)                                       |
| Benicar                    | Hypertension                      | [Olmesartan - Wikipedia](https://en.wikipedia.org/wiki/Olmesartan)                                       |
| Boniva                     | Osteoporosis                      | [Ibandronic acid - Wikipedia](https://en.wikipedia.org/wiki/Ibandronic_acid)                             |
| Bupropion SR               | Depression and smoking cessation  | [Bupropion - Wikipedia](https://en.wikipedia.org/wiki/Bupropion)                                         |
| Carisoprodol               | Muscle pain and discomfort        | [Carisoprodol - Wikipedia](https://en.wikipedia.org/wiki/Carisoprodol)                                   |
| Carvedilol                 | Heart failure and hypertension    | [Carvedilol - Wikipedia](https://en.wikipedia.org/wiki/Carvedilol)                                       |
| Cefdinir                   | Bacterial infections              | [Cefdinir - Wikipedia](https://en.wikipedia.org/wiki/Cefdinir)                                           |
| Celebrex                   | Pain and inflammation             | [Celecoxib - Wikipedia](https://en.wikipedia.org/wiki/Celecoxib)                                         |
| Celecoxib                  | Pain and inflammation             | [Celecoxib - Wikipedia](https://en.wikipedia.org/wiki/Celecoxib)                                         |
| Cephalexin                 | Bacterial infections              | [Cephalexin - Wikipedia](https://en.wikipedia.org/wiki/Cephalexin)                                       |
| Cialis                     | Erectile dysfunction              | [Tadalafil - Wikipedia](https://en.wikipedia.org/wiki/Tadalafil)                                         |
| Ciprofloxacin              | Bacterial infections              | [Ciprofloxacin - Wikipedia](https://en.wikipedia.org/wiki/Ciprofloxacin)                                 |
| Citalopram                 | Depression and anxiety            | [Citalopram - Wikipedia](https://en.wikipedia.org/wiki/Citalopram)                                       |
| Clavulanate K+             | Antibiotic                        | [Clavulanic acid - Wikipedia](https://en.wikipedia.org/wiki/Clavulanic_acid)                             |
| Clonazepam                 | Seizures and panic disorders      | [Clonazepam - Wikipedia](https://en.wikipedia.org/wiki/Clonazepam)                                       |
| Clonidine HCI              | Hypertension                      | [Clonidine - Wikipedia](https://en.wikipedia.org/wiki/Clonidine)                                         |
| Clopidogrel                | Blood clot prevention             | [Clopidogrel - Wikipedia](https://en.wikipedia.org/wiki/Clopidogrel)                                     |
| Clotrimazole               | Fungal infections                 | [Clotrimazole - Wikipedia](https://en.wikipedia.org/wiki/Clotrimazole)                                   |
| Codeine                    | Pain relief                       | [Codeine - Wikipedia](https://en.wikipedia.org/wiki/Codeine)                                             |
| Crestor                    | High cholesterol                  | [Rosuvastatin - Wikipedia](https://en.wikipedia.org/wiki/Rosuvastatin)                                   |
| Cyclobenzaprine            | Muscle pain and discomfort        | [Cyclobenzaprine - Wikipedia](https://en.wikipedia.org/wiki/Cyclobenzaprine)                             |
| Cymbalta                   | Depression and pain relief        | [Duloxetine - Wikipedia](https://en.wikipedia.org/wiki/Duloxetine)                                       |
| Dextroamphetamine XR       | ADHD                              | [Amphetamine/dextroamphetamine - Wikipedia](https://en.wikipedia.org/wiki/Amphetamine/dextroamphetamine) |
| Diazepam                   | Anxiety and seizures              | [Diazepam - Wikipedia](https://en.wikipedia.org/wiki/Diazepam)                                           |
| Diclofenac Sodium          | Pain and inflammation             | [Diclofenac - Wikipedia](https://en.wikipedia.org/wiki/Diclofenac)                                       |
| Doxycycline Hyclate        | Bacterial infections              | [Doxycycline - Wikipedia](https://en.wikipedia.org/wiki/Doxycycline)                                     |
| Duloxetine                 | Depression and pain relief        | [Duloxetine - Wikipedia](https://en.wikipedia.org/wiki/Duloxetine)                                       |
| Enalapril                  | Hypertension                      | [Enalapril - Wikipedia](https://en.wikipedia.org/wiki/Enalapril)                                         |
| Escitalopram               | Depression and anxiety            | [Escitalopram - Wikipedia](https://en.wikipedia.org/wiki/Escitalopram)                                   |
| Esomeprazole               | Acid reflux and ulcers            | [Esomeprazole - Wikipedia](https://en.wikipedia.org/wiki/Esomeprazole)                                   |
| Ezetimibe                  | High cholesterol                  | [Ezetimibe - Wikipedia](https://en.wikipedia.org/wiki/Ezetimibe)                                         |
| Fenofibrate                | High cholesterol                  | [Fenofibrate - Wikipedia](https://en.wikipedia.org/wiki/Fenofibrate)                                     |
| Fexofenadine               | Allergies                         | [Fexofenadine - Wikipedia](https://en.wikipedia.org/wiki/Fexofenadine)                                   |
| Finasteride                | Enlarged prostate and hair loss   | [Finasteride - Wikipedia](https://en.wikipedia.org/wiki/Finasteride)                                     |
| Flovent HFA 110mcg Inhaler | Asthma                            | [Fluticasone - Wikipedia](https://en.wikipedia.org/wiki/Fluticasone)                                     |
| Fluconazole                | Fungal infections                 | [Fluconazole - Wikipedia](https://en.wikipedia.org/wiki/Fluconazole)                                     |
| Fluoxetine HCI             | Depression and anxiety            | [Fluoxetine - Wikipedia](https://en.wikipedia.org/wiki/Fluoxetine)                                       |
| Fluticasone                | Asthma and allergies              | [Fluticasone - Wikipedia](https://en.wikipedia.org/wiki/Fluticasone)                                     |
| Fluticasone Nasal Spray    | Allergic rhinitis                 | [Fluticasone - Wikipedia](https://en.wikipedia.org/wiki/Fluticasone)                                     |
| Folic Acid                 | Vitamin deficiency                | [Folic acid - Wikipedia](https://en.wikipedia.org/wiki/Folic_acid)                                       |
| Furosemide                 | Edema and hypertension            | [Furosemide - Wikipedia](https://en.wikipedia.org/wiki/Furosemide)                                       |
| Gabapentin                 | Seizures and nerve pain           | [Gabapentin - Wikipedia](https://en.wikipedia.org/wiki/Gabapentin)                                       |
| Glimepiride                | Type 2 diabetes                   | [Glimepiride - Wikipedia](https://en.wikipedia.org/wiki/Glimepiride)                                     |
| Glipizide                  | Type 2 diabetes                   | [Glipizide - Wikipedia](https://en.wikipedia.org/wiki/Glipizide)                                         |
| Glyburide                  | Type 2 diabetes                   | [Glyburide - Wikipedia](https://en.wikipedia.org/wiki/Glyburide)                                         |
| Hydrochlorothiazide        | Hypertension and edema            | [Hydrochlorothiazide - Wikipedia](https://en.wikipedia.org/wiki/Hydrochlorothiazide)                     |
| Hydrocodone                | Pain relief and cough suppression | [Hydrocodone - Wikipedia](https://en.wikipedia.org/wiki/Hydrocodone)                                     |
| Hydrocortisone 2.5% Cream  | Skin inflammation                 | [Hydrocortisone - Wikipedia](https://en.wikipedia.org/wiki/Hydrocortisone)                               |
| Ibuprofen                  | Pain and inflammation             | [Ibuprofen - Wikipedia](https://en.wikipedia.org/wiki/Ibuprofen)                                         |
| Isosorbide Mononitrate     | Angina                            | [Isosorbide mononitrate - Wikipedia](https://en.wikipedia.org/wiki/Isosorbide_mononitrate)               |
| Lansoprazole               | Acid reflux and ulcers            | [Lansoprazole - Wikipedia](https://en.wikipedia.org/wiki/Lansoprazole)                                   |
| Lantus                     | Diabetes                          | [Insulin glargine - Wikipedia](https://en.wikipedia.org/wiki/Insulin_glargine)                           |
| Levofloxacin               | Bacterial infections              | [Levofloxacin - Wikipedia](https://en.wikipedia.org/wiki/Levofloxacin)                                   |
| Levothyroxine Sodium       | Hypothyroidism                    | [Levothyroxine - Wikipedia](https://en.wikipedia.org/wiki/Levothyroxine)                                 |
| Lisinopril                 | Hypertension                      | [Lisinopril - Wikipedia](https://en.wikipedia.org/wiki/Lisinopril)                                       |
| Lorazepam                  | Anxiety and insomnia              | [Lorazepam - Wikipedia](https://en.wikipedia.org/wiki/Lorazepam)                                         |
| Losartan                   | Hypertension                      | [Losartan - Wikipedia](https://en.wikipedia.org/wiki/Losartan)                                           |
| Lovastatin                 | High cholesterol                  | [Lovastatin - Wikipedia](https://en.wikipedia.org/wiki/Lovastatin)                                       |
| Meloxicam                  | Pain and inflammation             | [Meloxicam - Wikipedia](https://en.wikipedia.org/wiki/Meloxicam)                                         |
| Metformin                  | Type 2 diabetes                   | [Metformin - Wikipedia](https://en.wikipedia.org/wiki/Metformin)                                         |
| Metformin HCI              | Type 2 diabetes                   | [Metformin - Wikipedia](https://en.wikipedia.org/wiki/Metformin)                                         |
| Methylprednisone           | Inflammatory conditions           | [Methylprednisolone - Wikipedia](https://en.wikipedia.org/wiki/Methylprednisolone)                       |
| Metoprolol                 | Hypertension and angina           | [Metoprolol - Wikipedia](https://en.wikipedia.org/wiki/Metoprolol)                                       |
| Metoprolol Succinate XL    | Hypertension                      | [Metoprolol succinate - Wikipedia](https://en.wikipedia.org/wiki/Metoprolol_succinate)                   |
| Metoprolol Tartrate        | Hypertension                      | [Metoprolol tartrate - Wikipedia](https://en.wikipedia.org/wiki/Metoprolol_tartrate)                     |
| Mometasone                 | Inflammatory conditions           | [Mometasone - Wikipedia](https://en.wikipedia.org/wiki/Mometasone)                                       |
| Naproxen                   | Pain and inflammation             | [Naproxen - Wikipedia](https://en.wikipedia.org/wiki/Naproxen)                                           |
| Omeprazole                 | Acid reflux and ulcers            | [Omeprazole - Wikipedia](https://en.wikipedia.org/wiki/Omeprazole)                                       |
| Oxycodone                  | Pain relief                       | [Oxycodone - Wikipedia](https://en.wikipedia.org/wiki/Oxycodone)                                         |
| Pantoprazole               | Acid reflux and ulcers            | [Pantoprazole - Wikipedia](https://en.wikipedia.org/wiki/Pantoprazole)                                   |
| Paroxetine                 | Depression and anxiety            | [Paroxetine - Wikipedia](https://en.wikipedia.org/wiki/Paroxetine)                                       |
| Pioglitazone               | Type 2 diabetes                   | [Pioglitazone - Wikipedia](https://en.wikipedia.org/wiki/Pioglitazone)                                   |
| Potassium Chloride         | Electrolyte imbalance             | [Potassium chloride - Wikipedia](https://en.wikipedia.org/wiki/Potassium_chloride)                       |
| Pravastatin                | High cholesterol                  | [Pravastatin - Wikipedia](https://en.wikipedia.org/wiki/Pravastatin)                                     |
| Prednisone                 | Inflammatory conditions           | [Prednisone - Wikipedia](https://en.wikipedia.org/wiki/Prednisone)                                       |
| Pregabalin                 | Neuropathic pain and seizures     | [Pregabalin - Wikipedia](https://en.wikipedia.org/wiki/Pregabalin)                                       |
| Premarin                   | Hormone replacement therapy       | [Conjugated estrogens - Wikipedia](https://en.wikipedia.org/wiki/Conjugated_estrogens)                   |
| Promethazine               | Allergies and nausea              | [Promethazine - Wikipedia](https://en.wikipedia.org/wiki/Promethazine)                                   |
| Quetiapine                 | Mental health disorders           | [Quetiapine - Wikipedia](https://en.wikipedia.org/wiki/Quetiapine)                                       |
| Ranitidine                 | Acid reflux                       | [Ranitidine - Wikipedia](https://en.wikipedia.org/wiki/Ranitidine)                                       |
| Rosuvastatin               | High cholesterol                  | [Rosuvastatin - Wikipedia](https://en.wikipedia.org/wiki/Rosuvastatin)                                   |
| Salmeterol Inhaler         | Asthma and COPD                   | [Salmeterol - Wikipedia](https://en.wikipedia.org/wiki/Salmeterol)                                       |
| Sertraline HCI             | Depression and anxiety            | [Sertraline - Wikipedia](https://en.wikipedia.org/wiki/Sertraline)                                       |
| Simvastatin                | High cholesterol                  | [Simvastatin - Wikipedia](https://en.wikipedia.org/wiki/Simvastatin)                                     |
| Spironolactone             | Hypertension and edema            | [Spironolactone - Wikipedia](https://en.wikipedia.org/wiki/Spironolactone)                               |
| Sulfamethoxazole           | Bacterial infections              | [Sulfamethoxazole - Wikipedia](https://en.wikipedia.org/wiki/Sulfamethoxazole)                           |
| Synthroid                  | Hypothyroidism                    | [Levothyroxine - Wikipedia](https://en.wikipedia.org/wiki/Levothyroxine)                                 |
| Tamsulosin                 | Enlarged prostate                 | [Tamsulosin - Wikipedia](https://en.wikipedia.org/wiki/Tamsulosin)                                       |
| Temazepam                  | Insomnia                          | [Temazepam - Wikipedia](https://en.wikipedia.org/wiki/Temazepam)                                         |
| Topiramate                 | Epilepsy and migraines            | [Topiramate - Wikipedia](https://en.wikipedia.org/wiki/Topiramate)                                       |
| Tramadol                   | Pain relief                       | [Tramadol - Wikipedia](https://en.wikipedia.org/wiki/Tramadol)                                           |
| Trazodone HCI              | Depression and insomnia           | [Trazodone - Wikipedia](https://en.wikipedia.org/wiki/Trazodone)                                         |
| Triamcinolone Ace Topical  | Skin inflammation                 | [Triamcinolone acetonide - Wikipedia](https://en.wikipedia.org/wiki/Triamcinolone_acetonide)             |
| Triamterene                | Hypertension and edema            | [Triamterene - Wikipedia](https://en.wikipedia.org/wiki/Triamterene)                                     |
| Trimethoprim DS            | Bacterial infections              | [Trimethoprim/sulfamethoxazole - Wikipedia](https://en.wikipedia.org/wiki/Trimethoprim/sulfamethoxazole) |
| Valaciclovir               | Herpes simplex and shingles       | [Valaciclovir - Wikipedia](https://en.wikipedia.org/wiki/Valaciclovir)                                   |
| Valsartan                  | Hypertension                      | [Valsartan - Wikipedia](https://en.wikipedia.org/wiki/Valsartan)                                         |
| Venlafaxine XR             | Depression and anxiety            | [Venlafaxine - Wikipedia](https://en.wikipedia.org/wiki/Venlafaxine)                                     |
| Verapamil SR               | Hypertension and angina           | [Verapamil - Wikipedia](https://en.wikipedia.org/wiki/Verapamil)                                         |
| Viagra                     | Erectile dysfunction              | [Sildenafil - Wikipedia](https://en.wikipedia.org/wiki/Sildenafil)                                       |
| Yaz                        | Birth control                     | [Drospirenone/ethinylestradiol - Wikipedia](https://en.wikipedia.org/wiki/Drospirenone/ethinylestradiol) |
| Zolpidem                   | Insomnia                          | [Zolpidem - Wikipedia](https://en.wikipedia.org/wiki/Zolpidem)                                           |

<a id="research-question"></a>
# A1: Research Question

#### Scenario 2:

>One of the most critical factors in patient relationship management that directly affects a hospital’s long-term cost effectiveness is understanding its patients and the conditions leading to hospital admissions. When a hospital can better understand its patients’ characteristics, it is better able to target treatment to patients, resulting in more effective cost of care for the hospital in the long term.
>You are an analyst for a hospital that wants to better understand the characteristics of its patients. You have been asked to perform a market basket analysis to analyze patient data to identify key associations of your patients, ultimately allowing better business and strategic decision-making for the hospital.

#### Research Question:
Which medications should be grouped together to provide a more efficient layout in the Pharmacy for patient medication dispensing?


<a id="objectives"></a>
# A2: Objectives and Goals of Analysis

The goal of this Data Mining Report through Market Basket Analysis is to reveal essential correlations among variables within the dataset, specifically related to medication purchases. A question that resonates with real-world healthcare and pharmaceutical needs is the exploration of relationships between these variables, leading to the recognition of current medication purchasing patterns that may have been previously overlooked. The primary objective is to investigate these connections, evaluating how they can support the organization in optimizing medication sales and distribution, the findings of which will be communicated to stakeholders.

<a id="explination-of-market-basket-analysis"></a>
# B1: Explanation of Market Basket Analysis

Market Basket Analysis functions on datasets composed solely of binary values, either True or False, or 1's and 0's, distinguishing it from models that utilize continuous or ordinal data. This method is centered on finding recurring combinations of items in various transactions, enabling retailers to discern patterns in the products that customers often purchase in tandem. The anticipated results of this analysis are generally split into two parts known as Antecedents (IFs) and Consequents (THENs). These terms form the basis for conditional rules or scenarios, such as: If a customer bought a Nintendo game, they also purchased a Nintendo accessory. A fundamental concept of Market Basket Analysis, especially when employing the Apriori algorithm, is that any portion of a frequently observed itemset must itself be frequent.

<a id="transaction-example"></a>
# B2: Transaction Example

In [171]:
from typing import List, Tuple
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from mlxtend.frequent_patterns import association_rules, apriori
from mlxtend.preprocessing import TransactionEncoder

medical_market_basket_data = pd.read_csv('./medical_market_basket.csv')

#data-cleaning, remove empty rows.
medical_market_basket_data.dropna(how='all', inplace=True)
medical_market_basket_data.reset_index(inplace=True)
medical_market_basket_data.drop(columns=['index'], inplace=True)

'''
Transaction data example:

 1. Each row represents a patient transaction. 
 2. Each Column Presc# (1 - 20) represents a single item/prescription ordered.
'''
medical_market_basket_data.loc[4]

Presc01         abilify
Presc02    atorvastatin
Presc03      folic acid
Presc04        naproxen
Presc05        losartan
Presc06             NaN
Presc07             NaN
Presc08             NaN
Presc09             NaN
Presc10             NaN
Presc11             NaN
Presc12             NaN
Presc13             NaN
Presc14             NaN
Presc15             NaN
Presc16             NaN
Presc17             NaN
Presc18             NaN
Presc19             NaN
Presc20             NaN
Name: 4, dtype: object

Looking at this example transaction from the `WGU Medical Market Basket` Dataset we can see this patients transaction includes 5 items out of the allowable 20.

This transaction includes abilify, atorvastatin, folic acid, naproxen, losartan.

<a id="assumptions"></a>
# B3: Assumptions of Market Basket Analysis 

###### The submission adequately summarizes 1 assumption of market basket analysis.

The fundamental premise in market basket analysis revolves around the notion that when two or more products consistently appear together in various baskets, it signifies that these products are interconnected in purchases. Consequently, acquiring one product is likely to result in the purchase of others. Within this investigation, our focus centers on three-item rules.

<a id="ata-prepartion"></a>
# C1: Data Preparation

The raw `medical_market_basket.csv` was loaded into a [`Pandas DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) in *Section B2*. It also did a basic clean-up step to remove the blank rows between each actual transaction of the dataset.

The data must be in a jagged array or list of lists format to be processed by the [`TransactionEncoder`](https://rasbt.github.io/mlxtend/user_guide/preprocessing/TransactionEncoder/#transactionencoder-convert-item-lists-into-transaction-data-for-frequent-itemset-mining). Here we will create and execute a Python function that will pre-process the raw dataset for the Basket Market Analysis pre-processor.

In [172]:
'''
Name: convert_for_transaction_encoder
Parameters: Data as a Pandas DataFrame
Returns: Transactions as a jagged list
'''
def convert_for_transaction_encoder(data: pd.DataFrame) -> List[List[str]]:
    transactions = list()
    for row in range(len(data)):
        current_transaction = list()
        for column in range(len(data.columns)):
            current_value = data.iloc[row, column]
            if not pd.isna(current_value): #no NaN allowed
                current_transaction.append(str(current_value))
        transactions.append(current_transaction)
        
    return transactions

transaction_data = convert_for_transaction_encoder(medical_market_basket_data)
print(f'Total Transactions: [{len(transaction_data)}]')
    

Total Transactions: [7501]


In [173]:
'''
Check a few transactions to make sure they are in list format correctly.
'''
middle = int((len(transaction_data) / 2) - 1)
last = len(transaction_data) - 1

print(f'First Transaction: {transaction_data[0]}')
print(f'Middle Transaction: {transaction_data[middle]}')
print(f'Last Transaction: {transaction_data[last]}')

First Transaction: ['amlodipine', 'albuterol aerosol', 'allopurinol', 'pantoprazole', 'lorazepam', 'omeprazole', 'mometasone', 'fluconozole', 'gabapentin', 'pravastatin', 'cialis', 'losartan', 'metoprolol succinate XL', 'sulfamethoxazole', 'abilify', 'spironolactone', 'albuterol HFA', 'levofloxacin', 'promethazine', 'glipizide']
Middle Transaction: ['carvedilol', 'abilify', 'levofloxacin', 'doxycycline hyclate']
Last Transaction: ['amphetamine salt combo xr', 'levofloxacin', 'diclofenac sodium', 'cialis']


In [174]:
'''
Perform Transaction Encoding for Market Basket Analysis
'''

transaction_encoder = TransactionEncoder()

#fit the data to the encoder
transaction_encoder.fit(transaction_data)

#transform the data after fitting
transformed_transaction = transaction_encoder.transform(transaction_data)

cleaned_encoded_transactions = pd.DataFrame(transformed_transaction, columns=transaction_encoder.columns_)

cleaned_encoded_transactions

Duloxetine  Premarin    Yaz  abilify  acetaminophen  actonel  \
0          False     False  False     True          False    False   
1          False     False  False    False          False    False   
2          False     False  False    False          False    False   
3          False     False  False    False          False    False   
4          False     False  False     True          False    False   
...          ...       ...    ...      ...            ...      ...   
7496       False     False  False    False          False    False   
7497       False     False  False    False          False    False   
7498       False     False  False    False          False    False   
7499       False     False  False    False          False    False   
7500       False     False  False    False          False    False   

      albuterol HFA  albuterol aerosol  alendronate  allopurinol  ...  \
0              True               True        False         True  ...   
1             False              False        False        False  ...   
2             False              False        False        False  ...   
3             False              False        False         True  ...   
4             False              False        False        False  ...   
...             ...                ...          ...          ...  ...   
7496          False              False        False        False  ...   
7497          False              False        False        False  ...   
7498          False              False        False        False  ...   
7499          False              False        False        False  ...   
7500          False              False        False        False  ...   

      trazodone HCI  triamcinolone Ace topical  triamterene  trimethoprim DS  \
0             False                      False        False            False   
1             False                      False        False            False   
2             False                      False        False            False   
3             False                      False        False            False   
4             False                      False        False            False   
...             ...                        ...          ...              ...   
7496          False                      False        False            False   
7497          False                      False        False            False   
7498          False                      False        False            False   
7499          False                      False        False            False   
7500          False                      False        False            False   

      valaciclovir  valsartan  venlafaxine XR  verapamil SR  viagra  zolpidem  
0            False      False           False         False   False     False  
1            False      False           False         False   False     False  
2            False      False           False         False   False     False  
3            False      False           False         False   False     False  
4            False      False           False         False   False     False  
...            ...        ...             ...           ...     ...       ...  
7496         False      False           False         False   False     False  
7497         False      False           False         False   False     False  
7498         False      False           False         False   False     False  
7499         False      False           False         False   False     False  
7500         False      False           False         False   False     False  

[7501 rows x 119 columns]

In [175]:
'''
Save Encoded Transactions to CSV
'''

cleaned_encoded_transactions.to_csv('./task3-encoded-transactions.csv', index=False)

<a id="geneartion-of-association-rules"></a>
# C2: Generation of Association Rules

The process of generating association rules involves the utilization of two distinct functions. Initially, the [`apriori`](https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/apriori/) function is employed to extract all associations that satisfy a specified minimum support threshold. 

Subsequently, the [`association_rules`](https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/) function is employed to elaborate on these rules, incorporating an additional level of reduction by imposing yet another minimum threshold. This secondary threshold pertains to the lift value of each rule, and it serves as a focal point for fine-tuning the rule selection.


In [176]:
'''
Execute the "Apriori" method to get the frequencies (support value) of Itemsets.
'''
itemsets_support_frequencies = apriori(cleaned_encoded_transactions, min_support = 0.02, use_colnames = True)

#display results
itemsets_support_frequencies

support                          itemsets
0    0.046794                        (Premarin)
1    0.238368                         (abilify)
2    0.020397               (albuterol aerosol)
3    0.033329                     (allopurinol)
4    0.079323                      (alprazolam)
..        ...                               ...
98   0.023064            (lisinopril, diazepam)
99   0.023464              (diazepam, losartan)
100  0.022930            (diazepam, metoprolol)
101  0.020131  (glyburide, doxycycline hyclate)
102  0.028530             (glyburide, losartan)

[103 rows x 2 columns]

In [177]:
'''
Execute 'association_rules' function with 'Lift' metric and minimum threshold of 1.

With a threshold of 1 while executing 'association_rules' the odds of receiving antecedents (IFs) with Consequent (THENs) increases.
'''
rules = association_rules(itemsets_support_frequencies, metric = 'lift', min_threshold = 1.0)

print(f'Total Number of Rules Generated: [{len(rules)}]')

#rules

Total Number of Rules Generated: [94]


#### Summary of `Min_Support` and `Min_Threshold` value choices:

###### *Min_Support*
A higher minimum support value increases the exclusivity of the generated rules, while a lower value might include rules with limited significance due to small sample sizes. Using a minimum support value of 0.02 (2%) strikes a balance between inclusivity and restrictiveness

<br />

###### *Min_Threshold*
Lift measures the strength of association between the antecedent and the consequent of a rule. A lift of 1.0 indicates no association, values greater than 1.0 signify positive association, and values below 1.0 imply negative association. By setting the minimum lift to 1.0, the metric focuses on rules where the antecedent indeed increases the likelihood of the consequent appearing in transactions.

<a id="association-rules-table"></a>
# C3: Association Rules Table
###### Display Rules Table

In [178]:
rules

antecedents                  consequents  antecedent support  \
0                  (abilify)                 (amlodipine)            0.238368   
1               (amlodipine)                    (abilify)            0.071457   
2   (amphetamine salt combo)                    (abilify)            0.068391   
3                  (abilify)     (amphetamine salt combo)            0.238368   
4                  (abilify)  (amphetamine salt combo xr)            0.238368   
..                       ...                          ...                 ...   
89              (metoprolol)                   (diazepam)            0.095321   
90               (glyburide)        (doxycycline hyclate)            0.170911   
91     (doxycycline hyclate)                  (glyburide)            0.095054   
92               (glyburide)                   (losartan)            0.170911   
93                (losartan)                  (glyburide)            0.132116   

    consequent support   support  confidence      lift  leverage  conviction  \
0             0.071457  0.023597    0.098993  1.385352  0.006564    1.030562   
1             0.238368  0.023597    0.330224  1.385352  0.006564    1.137144   
2             0.238368  0.024397    0.356725  1.496530  0.008095    1.183991   
3             0.068391  0.024397    0.102349  1.496530  0.008095    1.037830   
4             0.179709  0.050927    0.213647  1.188845  0.008090    1.043158   
..                 ...       ...         ...       ...       ...         ...   
89            0.163845  0.022930    0.240559  1.468215  0.007312    1.101015   
90            0.095054  0.020131    0.117785  1.239135  0.003885    1.025766   
91            0.170911  0.020131    0.211781  1.239135  0.003885    1.051852   
92            0.132116  0.028530    0.166927  1.263488  0.005950    1.041786   
93            0.170911  0.028530    0.215943  1.263488  0.005950    1.057436   

    zhangs_metric  
0        0.365218  
1        0.299568  
2        0.356144  
3        0.435627  
4        0.208562  
..            ...  
89       0.352502  
90       0.232768  
91       0.213256  
92       0.251529  
93       0.240286  

[94 rows x 10 columns]

<a id="top-rules"></a>
# C4: Top Rules

In [179]:
selection = (rules['confidence'] > 0.29) & (rules['lift'] > 1.9)
rules_sorted_by_lift = rules[selection].copy().sort_values(by=['lift'], ascending= False)
top_3_rules = rules_sorted_by_lift.head(3).copy()

top_3_rules

antecedents   consequents  antecedent support  consequent support  \
74  (lisinopril)  (carvedilol)            0.098254            0.174110   
73   (glipizide)  (carvedilol)            0.065858            0.174110   
31   (metformin)     (abilify)            0.050527            0.238368   

     support  confidence      lift  leverage  conviction  zhangs_metric  
74  0.039195    0.398915  2.291162  0.022088    1.373997       0.624943  
73  0.022930    0.348178  1.999758  0.011464    1.267048       0.535186  
31  0.023064    0.456464  1.914955  0.011020    1.401255       0.503221

The higher lift indicates a stronger effect of the consequent being included in the transaction with the antecedent. The confidence score indicates a proportion of all transactions including the antecedent also featuring the consequent, and imposing a moderate confidence threshold here requires that we see a more significant proportion of antecedent transactions also include the consequent.

I initially set the `Confidence: 0.4` and `Lift: 2.0`, but this did not yield any results. I then gradually reduced these numbers until I obtained a set of rules that were worth sorting and selecting the top three from. The final values I settled on were `Confidence: 0.20` and `Lift: 1.9`.

<a id="results-analysis"></a>
# D1: Results of Analysis

>The submission adequately summarizes the significance of support, lift, and confidence from the results of the analysis.

In [180]:
analysis_items = ['antecedents', 'consequents', 'support', 'lift', 'confidence']
analysis_values = top_3_rules[analysis_items]

analysis_values

antecedents   consequents   support      lift  confidence
74  (lisinopril)  (carvedilol)  0.039195  2.291162    0.398915
73   (glipizide)  (carvedilol)  0.022930  1.999758    0.348178
31   (metformin)     (abilify)  0.023064  1.914955    0.456464

### Analysis of Support, Lift, and Confidence of Top 3 Rules

1. **Rule 74 (lisinopril -> carvedilol):**
   - **Support:** 0.039195, indicating that the combination of lisinopril and carvedilol appears in approximately 3.92% of all transactions.
   - **Lift:** 2.291162, suggesting that carvedilol is 2.29 times more likely to be bought with lisinopril than on its own.
   - **Confidence:** 0.398915, meaning that 39.89% of transactions containing lisinopril also include carvedilol.

2. **Rule 73 (glipizide -> carvedilol):**
   - **Support:** 0.022930, indicating that the combination of glipizide and carvedilol appears in approximately 2.29% of all transactions.
   - **Lift:** 1.999758, suggesting that carvedilol is 2 times more likely to be bought with glipizide than on its own.
   - **Confidence:** 0.348178, meaning that 34.82% of transactions containing glipizide also include carvedilol.

3. **Rule 31 (metformin -> abilify):**
   - **Support:** 0.023064, indicating that the combination of metformin and abilify appears in approximately 2.31% of all transactions.
   - **Lift:** 1.914955, suggesting that abilify is 1.91 times more likely to be bought with metformin than on its own.
   - **Confidence:** 0.456464, meaning that 45.65% of transactions containing metformin also include abilify.

<a id="practical-significance"></a>
# D2: Practical Significance

<a id="recommended-action"></a>
# D3: Recommended Action

<a id="panopto-recording"></a>
# E: Panopto Recording

<a id="code-reference"></a>
# F: Code References

<a id="source-references"></a>
# G: Source References

 * Bruce, P., Bruce, A., &amp; Gedeck, P. (2020). Practical statistics for data scientists: 50+ essential concepts using r and python. O’Reilly Media. <br /> <br />
 * Geron, A. (2022). Hands-on machine learning with scikit-learn, Keras, and tensorflow 3E: Concepts, tools, and techniques to build Intelligent Systems. O’Reilly Media.   